DPL Analysis: make sure VLAs use only one basket

ktf opened the pull request:
Otherwise it seems to create issues with ROOT bulk reading if the baskets
are split across multiple clusters.

In particular, each row needs 4 bytes for the size of the elements in the row
and the _size branch needs to also have its size correctly set.

ktf commented:
@jgrosseo @aalkin I think we need at least this for the case where we have more than one element (on average) for VLA. 

I think actually the real issue is that mFieldSize is 0, however I do not have a quick way to reproduce this. Do we have anything on hyperloop / standalone?

aalkin commented on Framework/Core/src/TableTreeHelpers.cxx:
```diff
@@ -454,7 +454,7 @@ std::shared_ptr<TTree> TableToTree::process()
 
   for (auto& reader : mColumnReaders) {
-    int idealBasketSize = 1024 + reader->fieldSize() * mRows; // minimal additional size needed, otherwise we get 2 baskets
+    int idealBasketSize = 1024 + reader->fieldSize() * reader->columnEntries(); // minimal additional size needed, otherwise we get 2 baskets
```
Probably it would be better to set field size for the column correctly in the constructor. Currently it is a bare byte width of the element type, in case of a fixed list it is multiplied by the list size. For the variable-size list we can use `columnEntries / mRows` as an estimate of an average list size.

ktf replied:
What I do not understand is that fieldSize() should already be correct. I suspect that the small basket size I was witnessing was for collections where the number of items is much larger than then numbers of rows. That could happen for the indices, I guess...

jgrosseo commented:
@mfaggin This is about the read error which you spotted...

ktf committed:
DPL Analysis: make sure VLAs use only one basket

Otherwise it seems to create issues with ROOT bulk reading if the baskets
are split across multiple clusters.

In particular, each row needs 4 bytes for the size of the elements in the row
and the _size branch needs to also have its size correctly set.

ktf commented:
Ok, this new version is better. First of all, each VLA uses 4 bytes for the size AND there is the additional _size branch whose basket size needs to be set. There is still something I do not quite get, however this is for sure closer to the optimal case. Tested with sim_challenge and verifying the resulting trees have the expected number of baskets per branch.

alibuild commented:
Error while checking build/O2/fullCI for a9e420f12d038d55c09fe926c6043baa5f439e7c at 2024-09-20 06:20:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/13519/a9e420f12d038d55c09fe926c6043baa5f439e7c/build_O2_fullCI/pretty.html).

jgrosseo commented:
Btw. - and sorry for this comment at this stage - what would happen if we set it for all to 32k. We waste memory but not compressed space?

ktf commented:
As discussed offline, 32k is a minimum value, not a max one. As for over-provisioning the basket sizes, I did not do it because I feared it would increase the memory usage in writing / reading. As agreed, we can revaluate that later.

jgrosseo commented:
@chiarazampolli We have MCs which are affected by this and are corrupted. I think this fix should be ported to a release and then the MCs redone? I don't have a list what is all affected but I know that LHC24g3 is...

stefanopolitano commented:
Dear @jgrosseo, @chiarazampolli, all, for your information this issue was spotted also in the LHC24i2 production. I think it would be better to port this PR asap to avoid rerunning all the MC productions if not all the future MC productions will have this problem, won't they? Thanks!

chiarazampolli commented:
Removing  async-2024-PbPb-cpass0, since cpass0 is over.
